### PR TITLE
Un-Like Song Feature

### PR DESCRIPTION
**Pull request by h-le**
- Client-Side
  - [x] Add button/method to send `DELETE` request to server, with song data

- Server-Side
  - [x] Create route to `delete()` previously liked song in `likes` collection
  - [x] Simplified routes

Note:
- ~~I left a `TODO` for testing the `LikesComponent` component but I wanted to get your thoughts on my approach (primarily on the client-side)~~

### CODE REVIEW

File: src/app/likes/likes.component.spec.ts [hunk 1]
```diff
@@ -106,14 +108,16 @@ describe('LikesComponent', () => {
   beforeEach(() => {
     fixture = TestBed.createComponent(LikesComponent);
     component = fixture.componentInstance;
-    fixture.detectChanges();
+    /* fixture.detectChanges(); */
```
Comment (okkays) [inline src/app/likes/likes.component.spec.ts]: remove

File: src/app/likes/likes.component.html [hunk 2]
```diff
@@ -5,19 +5,33 @@
...
-    ><a
+  >
+    <mat-grid-tile
       *ngFor="let like of likes"
-      [attr.href]="like.url"
-      target="_blank"
-      title="{{like.title}} by {{like.artist}}"
+      colspan="1"
+      [style.background-image]="'url(' + like.song_art_image_url + ')' | safe: 'style'"
+      style="background-size: cover;"
```
Comment (okkays) [inline src/app/likes/likes.component.html]: prefer CSS to the style attribute (unless it's computed by angular as on the line above)

File: src/app/likes/likes.component.ts [hunk 3]
```diff
@@ -14,15 +14,48 @@ import {HttpService} from '../http.service';
...
+      ),
+      this.unliked$
+    ).pipe(
+      map(([likes, unliked]: [Like[], Like[]]) =>
+        likes.filter(
+          (like: Like) => !unliked.some((unlike: Like) => like.id === unlike.id)
+        )
+      )
+    );
+    this.hoveredLikeId = -1;
```
Comment (okkays) [inline src/app/likes/likes.component.ts]: `number|null` seems like the more canonical way to represent this in TS

File: app.py [hunk 4]
```diff
@@ -75,3 +80,14 @@ def post_like():
     song = flask.request.get_json()
     like = firebase.like_song(id_token, song)
     return flask.jsonify(like)
+
+
+@app.route('/_unlike', methods=['DELETE'])
```
Comment (okkays) [inline app.py]: FYI - you can simplify this routing a bit; routes are unique on the combination of url/method, not just url.  Given that, you could do something like:

```
@app.route('/_like/<id>', methods=['DELETE']) # Delete a like
@app.route('/_like', methods=['POST']) # Create a new like
@app.route('/_like', methods=['GET']) # Get a list of likes
```

Typical CRUD is:
```
@app.route('/things', methods=['GET']) # Get a list of things
@app.route('/things', methods=['POST']) # Create a thing
@app.route('/things/<id>', methods=['GET']) # Get a specific thing
@app.route('/things/<id>', methods=['PATCH']) # Update a specific thing
@app.route('/things/<id>', methods=['DELETE']) # Delete a specific thing
```

**Review by okkays — CHANGES_REQUESTED** [review]

File: src/app/likes/likes.component.spec.ts [hunk 5]
```diff
@@ -125,44 +127,129 @@ describe('LikesComponent', () => {
     });
   }));
 
+  it('x', async(() => {
```
Comment (okkays) [inline src/app/likes/likes.component.spec.ts]: it('x')

File: src/app/likes/likes.component.spec.ts [hunk 6]
```diff
@@ -125,44 +127,129 @@ describe('LikesComponent', () => {
...
+    fixture.debugElement
+      .queryAll(By.css('.mat-grid-tile'))[0]
+      .nativeElement.dispatchEvent(new Event('mouseover'));
+
+    fixture.detectChanges();
+
+    let unlikedIndex: number = 0;
+    const expectedUnliked: Like[][] = [[], [likes[0]]];
+
+    component.unliked$.subscribe(unliked => {
```
Comment (okkays) [inline src/app/likes/likes.component.spec.ts:142]: remember to check out catalyst async matchers when you get your access

File: src/app/likes/likes.component.spec.ts [hunk 7]
```diff
@@ -125,44 +127,129 @@ describe('LikesComponent', () => {
...
+          tile.nativeElement.getElementsByClassName('mat-grid-tile-header')[0]
+            .innerText
+        );
+        if (song_artist && song_artist.length == 3) {
+          return {
+            artist: String(song_artist[2]),
+            title: String(song_artist[1]),
+            url: String(tile.nativeElement.getElementsByTagName('a')[0].href),
+          };
+        }
```
Comment (okkays) [inline src/app/likes/likes.component.spec.ts]:
```suggestion
        if (!song_artist || song_artist.length !== 3) return;
        return {
          artist: String(song_artist[2]),
          title: String(song_artist[1]),
          url: String(tile.nativeElement.getElementsByTagName('a')[0].href),
        };
```

- Use guard clauses to reduce nesting
- Use === and similar for comparison, not ==

**Review by okkays — APPROVED** [review]

File: src/app/likes/likes.component.spec.ts (hunk 5 above)
Comment (h-le) [reply src/app/likes/likes.component.spec.ts]: ... 😅

File: src/app/likes/likes.component.spec.ts (hunk 6 above)
Comment (h-le) [reply src/app/likes/likes.component.spec.ts:142]: 👍

**Pull request merged**